### PR TITLE
Add a workaround for unbinding script members on Mono

### LINKED ISSUE #20355: Crash due to modified collection on actor destroyed
## System Information
- Operating System:  macOS 12.6
- OpenRA Version: bleed branch, commit 7f677f1
- OpenRA Mod: ra
- Source: self-compiled
- Mono version: 6.12.0.122

## Exception log

```
OpenRA engine version {DEV_VERSION}
Red Alert mod version {DEV_VERSION}
on map b884e9d4e8e59f89b0f338e32fdc52a0a51b6ae4 (Desert Shellmap by Scott_NZ).
Date: 2022-10-08 21:16:48Z
Operating System: OSX (Unix 21.6.0.0)
Runtime Version: Mono 6.12.0.122 (2020-02/c621c35ffa0 Wed Feb 10 00:51:43 EST 2021) CLR 4.0.30319.42000
Exception of type `System.InvalidOperationException`: Collection was modified; enumeration operation may not execute.
  at System.Collections.Generic.Dictionary`2+Enumerator[TKey,TValue].MoveNext () [0x00013] in <36514b14425c4337b446653c547aa9c3>:0 
  at OpenRA.Scripting.ScriptObjectWrapper.Unbind (System.Type targetType) [0x00042] in /Users/bburns/projects/OpenRA/OpenRA.Game/Scripting/ScriptObjectWrapper.cs:66 
  at OpenRA.Scripting.ScriptActorInterface.OnActorDestroyed () [0x0002c] in /Users/bburns/projects/OpenRA/OpenRA.Game/Scripting/ScriptActorInterface.cs:54 
  at OpenRA.Actor.<Dispose>b__90_0 (OpenRA.World w) [0x00065] in /Users/bburns/projects/OpenRA/OpenRA.Game/Actor.cs:423 
  at OpenRA.World.Tick () [0x001ba] in /Users/bburns/projects/OpenRA/OpenRA.Game/World.cs:449 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x000c4] in /Users/bburns/projects/OpenRA/OpenRA.Game/Game.cs:624 
  at OpenRA.Game.LogicTick () [0x00041] in /Users/bburns/projects/OpenRA/OpenRA.Game/Game.cs:648 
  at OpenRA.Game.Loop () [0x00129] in /Users/bburns/projects/OpenRA/OpenRA.Game/Game.cs:820 
  at OpenRA.Game.Run () [0x0003c] in /Users/bburns/projects/OpenRA/OpenRA.Game/Game.cs:873 
  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00010] in /Users/bburns/projects/OpenRA/OpenRA.Game/Game.cs:308 
  at OpenRA.Launcher.Program.Main (System.String[] args) [0x00046] in /Users/bburns/projects/OpenRA/OpenRA.Launcher/Program.cs:32 
```

## Additional information
Crash occurs not long after running the game. Sometimes it seems to run for longer than others before crashing, but I hadn't made it into an actual game session when any of the crashes occurred.

Not sure if it matters, but this is running on an M1 Pro.

### PR DESCRIPTION
**Pull request by abcdefg30**
Closes #20355.

### CODE REVIEW

**Comment by pchote** [comment]
I would expect something like this to work on both .NET and Mono:

```csharp
for (var i = members.Count; i >= 0; i--)
{
	var m = members[i];
	if (targetType == m.Value.Target.GetType())
		members.Remove(m.Key);
}
```

File: OpenRA.Game/Scripting/ScriptObjectWrapper.cs [hunk 1]
```diff
@@ -63,9 +67,22 @@ protected void Bind(object[] clrObjects)
...
+#else
+			// PERF: Re-use instead of allocating a new list on each unbind
+			membersToRemove.Clear();
+
+			foreach (var m in members)
+				if (targetType == m.Value.Target.GetType())
+					membersToRemove.Add(m.Key);
+
+			foreach (var m in membersToRemove)
+				members.Remove(m);
```
Comment (PunkPun) [inline OpenRA.Game/Scripting/ScriptObjectWrapper.cs:84]: You could use linq. members.Where.ToArray and then itterate over this array

**Comment by abcdefg30** [comment]
> I would expect something like this to work on both .NET and Mono:

Unfortunately it's not as simple to index a dictionary like this.

```
foreach (var key in members.Keys)
	if (targetType == members[key].Target.GetType())
		members.Remove(key);
```
came to mind as well, but I fear we'll run into the same problem, just with modifying the `KeyCollection` this time.

File: OpenRA.Game/Scripting/ScriptObjectWrapper.cs (hunk 1 above)
Comment (abcdefg30) [reply OpenRA.Game/Scripting/ScriptObjectWrapper.cs:84]: That would mean we'd still allocate a new array on each unbind. 🤔

**Review by PunkPun — APPROVED** [review]

**Review by RoosterDragon — APPROVED** [review]

**Comment by PunkPun** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/315dc8ce9ceb20515a004a544385f422cf9a5174...f6e902a974f7f7dbf2ae7f804b1c8f4685ebb692)

**Pull request merged**